supervisor/supervisor.go

```diff
@@ -0,0 +1,311 @@
+package supervisor
+
+import (
+	"fmt"
+	"github.com/google/uuid"
+	"github.com/pkg/errors"
+	"go.uber.org/zap"
+	"os/exec"
+	"sync"
+)
+
+var (
+	ErrJobNotFound  = errors.New("invalid job id")
+	ErrExecNotFound = errors.New("executable not found on system")
+)
+
+const (
+	StdOut = iota
+	StdErr
+)
+
+// LogOutput is a byte of output from a running job either
+// from standard output or stadard error
+type LogOutput struct {
+	Type uint8
+	Msg  byte
+}
+
+// Supervisor manages all processes
+type Supervisor struct {
+	access         *sync.RWMutex
+	jobs           *sync.Map
+	logger         *zap.Logger
+	ouputBuffers   map[string][]LogOutput
+	outputChannels map[string]*logWatchers
+	wg             *sync.WaitGroup
+}
+
+// NewSupervisor returns an instance of a Supervisor
+func NewSupervisor(logger *zap.Logger) *Supervisor {
+	return &Supervisor{
+		access:         &sync.RWMutex{},
+		jobs:           &sync.Map{},
+		logger:         logger,
+		ouputBuffers:   make(map[string][]LogOutput),
+		outputChannels: make(map[string]*logWatchers),
+		wg:             &sync.WaitGroup{},
+	}
+}
+
+// JobStatus is the status of a job with it's id
+type JobStatus struct {
+	Id     string
+	Status Status
+}
+
+// logWatchers is a wrapper to manage output streaming for each job.
+// It contains a set of chan LogOutput for each job.
+// WatchOutput adds the provided channel to this set so the caller can subscribe
+// to lines of output from the job
+type logWatchers struct {
+	*sync.RWMutex
+	watchers map[chan LogOutput]bool
+}
+
+// newLogWatcher returns an instance of a logWatchers
+func newLogWatcher() *logWatchers {
+	return &logWatchers{
+		RWMutex:  &sync.RWMutex{},
+		watchers: make(map[chan LogOutput]bool),
+	}
+}
+
+// addWatcher adds the provided channel to the logWatchers set
+func (l *logWatchers) addWatcher(c chan LogOutput) {
+	l.Lock()
+	defer l.Unlock()
+
+	l.watchers[c] = true
+}
+
+// sendLog sends the log message to all the subscribers in the logWatchers set
+func (l *logWatchers) sendLog(log LogOutput) {
+	l.RLock()
+	defer l.RUnlock()
+
+	for c := range l.watchers {
+		c <- log
+	}
+}
+
+// removeLogWatcher removes the channel from the logWatchers set
+func (l *logWatchers) removeLogWatcher(c chan LogOutput) {
+	l.Lock()
+	defer l.Unlock()
+
+	delete(l.watchers, c)
+}
+
+// Start starts a job and returns the current status. An ErrExecNotFound
+// is returned if the executable cannot be found by the system.
+// A unique id is generated for the job and returned in JobStatus
+func (s *Supervisor) Start(cmd string, args []string) (*JobStatus, error) {
+	if cmd == "" {
+		msg := "cannot start job, no executable defined"
+		s.logger.Error(msg)
+		return nil, errors.New(msg)
+	}
+
+	_, err := exec.LookPath(cmd)
```

Comment:
note: this would not be needed if `NewProcess` started the command right away (without a separate `Start` method) and returned the error from `p.cmd.Start` instead of using an internal `error` field

Reply:
Thanks. This has been fixed. The Process now starts right away in `NewProcess`